README.md
# `flatpak` cookbook

[![Cookbook Version](https://img.shields.io/cookbook/v/flatpak.svg)](https://supermarket.chef.io/cookbooks/flatpak)
[![CI State](https://github.com/detjensrobert/flatpak-cookbook/workflows/ci/badge.svg)](https://github.com/detjensrobert/flatpak-cookbook/actions?query=workflow%3Aci)
[![License](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT)

## Description

Chef cookbook for managing [Flatpak](https://github.com/flatpak/flatpak) applications.

## Notes

Currently, this cookbook only supports installing apps system-wide, and not per-user installs.

## Requirements

- Chef 15.3 or higher

## Platform

- CentOS 7+ (incl. Stream / Rocky / Alma)
- Amazon Linux 2
- Fedora
- Ubuntu 18.04+
- Debian 10+

## Resources

- [`flatpak_install`](documentation/install.md) - Set up Flatpak
- [`flatpak_remote`](documentation/remote.md) - Configure remotes
- [`flatpak_app`](documentation/app.md) - Install applications

## Minimal Example

```rb
flatpak_install 'example'

flatpak_app 'org.clementine_player.Clementine'
flatpak_app 'org.mozilla.firefox'
```

> **Note:** If the node is on a slow connection (e.g. < 1MB/s), the first `flatpak_app` install may time out and Chef will fail, as it will need to download the shared runtimes in addition to the application. If this happens, run Chef again and Flatpak will resume downloading where it left off.

## Contributing

PRs welcome!


CHANGELOG.md
# `flatpak` CHANGELOG

This file is used to list changes made in each version of the `flatpak` cookbook.

## 1.0.1 (7/17/2022)

- Update README badges
- Create GH release on tag push

## 1.0.0 (7/17/2022)

Initial release!

- `flatpak_install` to install flatpak
- `flatpak_remote` to configure remotes
- `flatpak_app` to manage applications
- Resource documentation
- Integration testing


test/integration/app/app_test.rb
control 'flatpak' do
  title 'Flatpak package is installed'

  describe package('flatpak') do
    it { should be_installed }
  end

  describe command('flatpak --version') do
    its('stdout') { should match /Flatpak 1\.\d+\.\d+/ }
  end
end

control 'app' do
  title 'Flatpak apps are installed'

  # describe command('flatpak list --columns application,origin') do
  describe command('flatpak list -d') do
    its('stdout') { should match /org.gnome.Weather.+flathub/ }
    its('stdout') { should match /org.gnome.Calculator.+fedora/ }
  end
end


test/integration/remote/remote_test.rb
control 'flatpak' do
  title 'Flatpak package is installed'

  describe package('flatpak') do
    it { should be_installed }
  end

  describe command('flatpak --version') do
    its('stdout') { should match /Flatpak 1\.\d+\.\d+/ }
  end
end

control 'remote' do
  title 'Flatpak remotes are configured correctly'

  describe file('/etc/flatpak/remotes.d/fedora.flatpakrepo') do
    it { should exist }
    its('content') { should match Regexp.escape('Url=oci+https://registry.fedoraproject.org') }
  end

  describe file('/etc/flatpak/remotes.d/fedora.filter') do
    it { should exist }
    its('content') { should match /deny \*/ }
    its('content') { should match /allow org.fedoraproject.\*/ }
  end

  describe file('/etc/flatpak/remotes.d/gnome-nightly.flatpakrepo') do
    it { should_not exist }
  end

  # describe command('flatpak remotes --columns name') do
  describe command('flatpak remotes') do
    its('stdout') { should match /fedora/ }
    its('stdout') { should_not match /gnome-nightly/ }
  end
end

control 'filter' do
  title 'Verify remote filter is honored correctly'
  only_if 'remote filtering not supported by old flatpak' do
    # only newer flatpak supports filters
    !((os.redhat? && os.release.to_i == 7) || os.name == 'amazon' || (os.debian? && os.release.to_i == 10))
    # command('flatpak --version').stdout.delete_prefix('Flatpak ').to_f >= 1.3
  end

  # recipe confgures remote to only show org.fedoraproject.*
  # describe command('flatpak remote-ls fedora --columns application') do
  describe command('flatpak remote-ls fedora') do
    its('stdout') { should match /org.fedoraproject.MediaWriter/ }
    its('stdout') { should_not match /org.gnome/ }
  end
end


test/cookbooks/flatpak-test/metadata.rb
name              'flatpak-test'
description       'Test cookbook for the flatpak cookbook'
version           '0.1.0'

depends 'flatpak'


resources/app.rb
provides :flatpak_app
unified_mode true

property :ref, String, name_property: true,
          description: 'Full ref of the application/runtime to install, possibly including arch/branch'

property :remote, String,
          description: 'Remote to install the application from'

property :flags, Array, default: [],
          coerce: proc { |p| p.sort },
          description: 'Additional CLI flags to specify when installing / updating'

action_class do
  include Flatpak::Cookbook::Helpers
end

action :install do
  # older flatpak versions need the remote to install from
  raise '`remote` must be specified for `flatpak_app` for flatpak versions < 1.1' unless new_resource.remote || flatpak_version >= 1.1

  unless flatpak_installed_apps.include? new_resource.ref
    converge_by "install #{new_resource.ref}#{(' from remote ' + new_resource.remote) if new_resource.remote}" do
      cmd = [
        "flatpak install #{noninteractive} --assumeyes",
        new_resource.remote,
        new_resource.ref,
        *new_resource.flags,
      ].join(' ')

      shell_out!(cmd, timeout: 3600)
    end
  end
end

action :update do
  if flatpak_installed_apps.include? new_resource.ref
    converge_by "update #{new_resource.ref}" do
      cmd = [
        "flatpak update #{noninteractive} --assumeyes",
        new_resource.ref,
        *new_resource.flags,
      ].join(' ')

      shell_out!(cmd, timeout: 3600)
    end
  end
end

action :remove do
  if flatpak_installed_apps.include? new_resource.ref
    converge_by "remove #{new_resource.ref}" do
      cmd = [
        "flatpak remove #{noninteractive} --assumeyes",
        new_resource.ref,
        *new_resource.flags,
      ].join(' ')

      shell_out!(cmd, timeout: 3600)
    end
  end
end


resources/remote.rb
provides :flatpak_remote
unified_mode true

property :remote_name, String, name_property: true,
          description: 'Name of the remote to configure'

# Either .location OR .url, etc are required
# .location is the url from `flatpak install URL`
# and contains all the other stuff
property :location, String,
          regex: /.+\.flatpakrepo/

property :url, String
property :title, String
property :comment, String
property :description, String
property :homepage, String
property :icon, String
property :gpgkey, String
property :collection_id, String

property :filters, Array,
          description: 'List of filters to apply, see flatpak-remote-add docs'

property :priority, Integer,
          coerce: proc { |p| p.to_i },
          description: 'Priority for the configured remote, higher will be chosen first'

action_class do
  include Flatpak::Cookbook::Helpers
end

action :add do
  # validate either of the required properties are set
  unless new_resource.location || new_resource.url
    raise "The #{new_resource.remote_name} `flatpak_remote` resource must have either .location or .url specified!"
  end
  # warn if both are set
  if new_resource.location && new_resource.url
    warn 'Preferring .location over manual .url, did you mean to set both?'
  end

  directory '/etc/flatpak/remotes.d' do
    recursive true
  end

  filter_file = nil
  if new_resource.filters
    warn 'Flatpak versions < 1.3.4 do not support filtering remotes' unless flatpak_supports_filters?

    filter_file = "/etc/flatpak/remotes.d/#{new_resource.remote_name}.filter"
    template filter_file do
      source 'remote.filter.erb'
      cookbook 'flatpak'
      variables(filters: new_resource.filters)
    end
  end

  if new_resource.location
    remote_file "/etc/flatpak/remotes.d/#{new_resource.remote_name}.flatpakrepo" do
      source new_resource.location
    end
  else
    template "/etc/flatpak/remotes.d/#{new_resource.remote_name}.flatpakrepo" do
      source 'remote.flatpakrepo.erb'
      cookbook 'flatpak'
      variables(
        name: new_resource.remote_name,
        url: new_resource.url,
        title: new_resource.title,
        comment: new_resource.comment,
        description: new_resource.description,
        homepage: new_resource.homepage,
        icon: new_resource.icon,
        gpgkey: new_resource.gpgkey,
        collection_id: new_resource.collection_id,
        filter_file: filter_file
      )
    end
  end

  # older versions of flatpak don't pick up flatpakrepo files in remotes.d
  # gotta tell it to import the remote manually
  unless flatpak_current_remotes.include?(new_resource.remote_name)
    converge_by "import /etc/flatpak/remotes.d/#{new_resource.remote_name}.flatpakrepo" do
      shell_out!("flatpak remote-add #{new_resource.remote_name} /etc/flatpak/remotes.d/#{new_resource.remote_name}.flatpakrepo")
    end
  end

  if new_resource.priority
    if flatpak_current_priorities[new_resource.remote_name].to_i != new_resource.priority.to_i
      converge_by "set priority for #{new_resource.remote_name}" do
        shell_out!("flatpak remote-modify #{new_resource.remote_name} --prio=#{new_resource.priority}")
      end
    end
  end
end

action :remove do
  file "/etc/flatpak/remotes.d/#{new_resource.remote_name}.flatpakrepo" do
    action :delete
  end

  file "/etc/flatpak/remotes.d/#{new_resource.remote_name}.filter" do
    action :delete
  end

  # older versions of flatpak don't update when the .flatpakrepo is deleted
  # gotta tell it to delete the remote manually
  if flatpak_current_remotes.include?(new_resource.remote_name)
    converge_by "remove #{new_resource.remote_name}" do
      shell_out!("flatpak remote-delete #{new_resource.remote_name}")
    end
  end
end


resources/remote_cli.rb
# provides :flatpak_remote
unified_mode true

property :remote_name, String, name_property: true,
          description: 'Name of the remote to configure'

property :location, String

property :url, String
property :title, String
property :comment, String
property :description, String
property :homepage, String
property :icon, String
property :gpgkey, String

property :priority, Integer,
          coerce: proc { |p| p.to_i },
          description: 'Priority for the configured remote, higher will be chosen first'

property :subset, String

# property :filter, Array, default: [],
#           description: 'List of filters to apply, see flatpak-remote-add docs'

property :flags, Array, default: [],
          coerce: proc { |p| p.sort },
          description: 'Additional CLI flags to specify when creating the repo'

# property :user, [true, false], default: false,
#           description: 'Use the per-user configuration instead of system-wide'

# podman output columns vs. flag names
# alias_method :url, :location

load_current_value do |new_resource|
  remotes = flatpak_get_remotes
  if remotes.key?(new_resource.remote_name)
    remotes[new_resource.remote_name].each do |key, value|
      next if key == 'options'
      send(key.to_sym, value)
    end
  else
    current_value_does_not_exist!
  end
end

action_class do
  include Flatpak::Cookbook::Helpers
end

action :add do
  converge_if_changed do
    remotes = flatpak_get_remotes

    cmd = [
      'flatpak',
      # (new_resource.user ? '--user' : '--system'),
      # modify existing remote, or create
      (remotes.key?(new_resource.remote_name) ? 'remote-modify' : 'remote-add'),
      # flatpak remote-add NAME LOCATION --OPTIONS
      new_resource.remote_name,
      new_resource.location,
      ("--prio='#{new_resource.priority}'" if new_resource.priority),
      ("--title='#{new_resource.title}'" if new_resource.title),
      ("--comment='#{new_resource.comment}'" if new_resource.comment),
      ("--description='#{new_resource.description}'" if new_resource.description),
      ("--homepage='#{new_resource.homepage}'" if new_resource.homepage),
      ("--icon='#{new_resource.icon}'" if new_resource.icon),
      ("--subset='#{new_resource.subset}'" if new_resource.subset),
      # ("--filter='#{new_resource.filter}'" if !new_resource.filter.empty?),
      *new_resource.flags, # any additional user-supplied ones
    ].join(' ')

    shell_out!(cmd)
  end
end

action :remove do
end


libraries/helpers.rb
module Flatpak
  module Cookbook
    module Helpers
      def flatpak_version
        # we can discard the patch version here
        # the checks we need are just major.minor for features
        # the main problem distros here are centos 7, debian 10, and amazon 2

        # we can save this as it wont change during the chef run
        @version ||= shell_out!('flatpak --version').stdout.delete_prefix('Flatpak ').to_f
      end

      def flatpak_supports_columns?
        flatpak_version >= 1.1
      end

      def flatpak_supports_filters?
        flatpak_version >= 1.3 # 1.3.4
      end

      def flatpak_current_remotes
        if flatpak_supports_columns?
          shell_out!('flatpak remotes --columns name').stdout.split
        else
          shell_out!('flatpak remotes').stdout.split("\n").map { |r| r.split("\t").first }
        end
      end

      def flatpak_current_priorities
        if flatpak_supports_columns?
          shell_out!('flatpak remotes --columns name,priority').stdout.split("\n").map(&:split).to_h
        else
          shell_out!('flatpak remotes -d').stdout.split("\n").map do |r|
            remote = r.split("\t")
            [remote[0], remote[-2]]
          end.to_h
        end
      end

      def flatpak_installed_apps
        if flatpak_supports_columns?
          shell_out!('flatpak list --columns application,ref').stdout.split
        else
          shell_out!('flatpak list -d').stdout.split("\n").map do |r|
            ref = r.split("\t").first # list on this version only returns the full ref
            [ref, ref.split('/').first] # so also extract the.application.id from it
          end.flatten
        end
      end

      def flatpak_updateable_apps
        if flatpak_supports_columns?
          shell_out!('flatpak remote-ls --updates --columns application,ref').stdout.split
        else
          shell_out!('flatpak remote-ls --updates -d').stdout.split("\n").map do |r|
            ref = r.split("\t").first # list on this version only returns the full ref
            [ref, ref.split('/').first] # so also extract the.application.id from it
          end.flatten
        end
      end

      def noninteractive
        '--noninteractive' if flatpak_version >= 1.2
      end
    end
  end
end
